torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -45,11 +46,11 @@ class DataLoader {
         this->worker_thread(std::move(dataset));
       });
     }
-    if (options_.workers == 0) {
+    if (options_.workers == 0 || options_.chunk_loading) {
```

Comment:
Thinking about this, I think there are two problems here:
1. If the main thread just keep a copy of the data set, the reset() method won't affect the data set in the working thread, i.e. the sampler in the working thread is still broken. 
2. Even we managed to have one copy of the data set and have everyone hold a reference to it, how do we keep the main thread and the worker thread in sync? For example, make the worker thread halt when we reset the data set in main thread?

Reply:
We make sure that when creating the chunk data loader using make_chunk_dataloader() we always wrap the Dataset that the user provided with the SharedDataSet. That points to a single reference no matter how many copies we make of the outside wrapper.

Reply:
I just look look at the SharedDataSet in https://github.com/pytorch/pytorch/pull/13800

It seems that data set is a shared_ptr to the real thing. If we are adopting that implementation, do we need to wrap it again with a unique pointer here? I am not sure if it's a good idea to wrap a shared pointer with a unique pointer.